addons/product_visible_discount/product_visible_discount.py

```diff
@@ -80,29 +101,36 @@ def get_real_price(res_dict, product_id, qty, uom, pricelist):
         result=res['value']
         pricelist_obj=self.pool.get('product.pricelist')
         product_obj = self.pool.get('product.product')
+
         if product and pricelist and self.pool.get('res.users').has_group(cr, uid, 'sale.group_discount_per_so_line'):
-            if result.get('price_unit',False):
+            if result.get('price_unit',False) is not False:
                 price=result['price_unit']
             else:
                 return res
             uom = result.get('product_uom', uom)
             product = product_obj.browse(cr, uid, product, context)
             pricelist_context = dict(context, uom=uom, date=date_order)
+
             list_price = pricelist_obj.price_rule_get(cr, uid, [pricelist],
                     product.id, qty or 1.0, partner_id, context=pricelist_context)
 
+            new_list_price, currency_id = get_real_price_curency(list_price, product.id, qty, uom, pricelist)
+
             so_pricelist = pricelist_obj.browse(cr, uid, pricelist, context=context)
 
-            new_list_price, currency_id = get_real_price_curency(list_price, product.id, qty, uom, pricelist)
-            if so_pricelist.visible_discount and list_price[pricelist][0] != 0 and new_list_price != 0:
+            if so_pricelist.visible_discount: # and list_price[pricelist][0] != 0 and new_list_price != 0:
+
                 if product.company_id and so_pricelist.currency_id.id != product.company_id.currency_id.id:
                     # new_list_price is in company's currency while price in pricelist currency
                     ctx = context.copy()
                     ctx['date'] = date_order
                     new_list_price = self.pool['res.currency'].compute(cr, uid,
-                        currency_id.id, so_pricelist.currency_id.id,
-                        new_list_price, context=ctx)
-                discount = (new_list_price - price) / new_list_price * 100
+                                                                       currency_id.id,
+                                                                       so_pricelist.currency_id.id,
+                                                                       new_list_price, context=ctx)
+
+                discount = (new_list_price - price) / new_list_price * 100 \
+                    if new_list_price else 100.0 if price else 0.0
```

Comment:
Some parentheses or putting the `if price` on top would be a bit more readable, IMHO.